plugin/core/sessions.py

```diff
@@ -277,6 +283,30 @@ def _handle_request_workspace_configuration(self, params: Dict[str, Any], reques
                 items.append(self.config.settings)
         self.client.send_response(Response(request_id, items))
 
+    def _handle_register_capability(self, params: Any, request_id: Any) -> None:
+        registrations = params["registrations"]
+        for registration in registrations:
+            method = registration["method"]
+            if method == "workspace/didChangeWorkspaceFolders":
+                self.capabilities.setdefault(
+                    "workspace", {}).setdefault(
+                        "workspaceFolders", {})["changeNotifications"] = registration["id"]
```

Comment:
The spec is hard to understand.

It says that `changeNotifications` can be statically set to an ID in which case server can use that ID later for `client/registerCapability`/`client/unregisterCapability`.

But it can also be a boolean in which case what? Is it then allowed to generate any ID for that call?

And what if that capability is not set? Should it then not be able to use dynamic registration?

We don't really do any validation of that logic but I guess it's fine for a start.

Reply:
Yeah. This is why I say: mutate the damn capabilities and be done with it :)
As long as it gets stuff working :)